internal/core/src/bitset/detail/element_wise.h

```diff
@@ -0,0 +1,976 @@
+            return;
+        }
+
+        // process the first element
+        if (start_shift != 0) {
+            const data_type existing_v = data[start_element];
+            const data_type new_v = ~existing_v;
+
+            const data_type existing_mask = get_shift_mask_begin(start_shift);
+            const data_type new_mask = get_shift_mask_end(start_shift);
+
+            data[start_element] =
+                (existing_v & existing_mask) | (new_v & new_mask);
+            start_element += 1;
+        }
+
+        // process the middle
+        for (size_type i = start_element; i < end_element; i++) {
+            data[i] = ~data[i];
+        }
+
+        // process the last element
+        if (end_shift != 0) {
+            const data_type existing_v = data[end_element];
+            const data_type new_v = ~existing_v;
+
+            const data_type existing_mask = get_shift_mask_end(end_shift);
+            const data_type new_mask = get_shift_mask_begin(end_shift);
+
+            data[end_element] =
+                (existing_v & existing_mask) | (new_v & new_mask);
+        }
+    }
+
+    static inline void
+    op_and(data_type* const left,
+           const data_type* const right,
+           const size_t start_left,
+           const size_t start_right,
+           const size_t size) {
+        op_func(left,
+                right,
+                start_left,
+                start_right,
+                size,
+                [](const data_type left_v, const data_type right_v) {
+                    return left_v & right_v;
+                });
+    }
+
+    static inline void
+    op_or(data_type* const left,
+          const data_type* const right,
+          const size_t start_left,
+          const size_t start_right,
+          const size_t size) {
+        op_func(left,
+                right,
+                start_left,
+                start_right,
+                size,
+                [](const data_type left_v, const data_type right_v) {
+                    return left_v | right_v;
+                });
+    }
+
+    static inline data_type
+    get_shift_mask_begin(const size_type shift) {
+        // 0 -> 0b00000000
+        // 1 -> 0b00000001
+        // 2 -> 0b00000011
+        if (shift == data_bits) {
+            return data_type(-1);
+        }
+
+        return (data_type(1) << shift) - data_type(1);
+    }
+
+    static inline data_type
+    get_shift_mask_end(const size_type shift) {
+        // 0 -> 0b11111111
+        // 1 -> 0b11111110
+        // 2 -> 0b11111100
+        return ~(get_shift_mask_begin(shift));
+    }
+
+    static inline void
+    set(data_type* const data, const size_type start, const size_type size) {
+        fill(data, start, size, true);
+    }
+
+    static inline void
+    reset(data_type* const data, const size_type start, const size_type size) {
+        fill(data, start, size, false);
+    }
+
+    static inline bool
+    all(const data_type* const data,
```

Comment:
all/none using a common pattern function like another op_func ?

Reply:
refactored that. All ops start from `op_`.